### ISSUE TITLE
TE: Redundant backward computation in PEFT setting.

### ISSUE THREAD

**Issue by kshitij12345** [Lightning-AI/lightning-thunder#2076]
Problem - Currently, TE executor just assumes that we will be computing gradients for both input and weight. This is true for full finetuning, but in PEFT scenario the weights are frozen i.e. requires_grad=False. This leads to poor performance as the TE executor ends up doing unnecessary work of computing gradients for weights.

NOTE: We tried fixing in #1838 but it had to be reverted due to #1966

**Comment by kshitij12345** [comment]
Conversation in https://github.com/Lightning-AI/lightning-thunder/issues/1966 discusses a potential solution 

> If instead of a single function which computes gradients for input, weight and bias (with FP8 GEMM), we could have a separate function (and symbol) to compute input, weight and bias gradients (need to check feasibility), then it should most likely work fine thanks to thunder's autograd routing the gradients correctly (and discarding computation when not required).

**Comment by riccardofelluga** [comment]
It looks like this can be done with the functional API 👀

https://github.com/NVIDIA/TransformerEngine/blob/1d903f5e6d5d36eef5f44bbffdc7719b703637e1/transformer_engine/pytorch/ops/basic/basic_linear.py#L580-L581

**Comment by riccardofelluga** [comment]
> Even for this new API, wgrad and dgrad are computed in a single function [_functional_backward](https://github.com/NVIDIA/TransformerEngine/blob/beaecf84ea93c817b0fe7e0611e49c6c0b2e6b30/transformer_engine/pytorch/ops/basic/basic_linear.py#L542-L622).

Why is there the need to split the computation if the _functional_backward provides toggle arguments to disable computing gradients for the weights? 

https://github.com/NVIDIA/TransformerEngine/blob/1d903f5e6d5d36eef5f44bbffdc7719b703637e1/transformer_engine/pytorch/ops/basic/basic_linear.py#L814-L845 

I think I am missing something

**Comment by riccardofelluga** [comment]
Hmmm interesting idea! I am also a fan of this #1599 which it a bit more elegant than a transform, however I don't know if it would solve all the problems. One plus point is that it is still an open PR so it would be up for changes, I think it's worth the investigation, using that + the gemm toggles from the functional API might do it 👀

**Comment by kshitij12345** [comment]
As far as I can tell #1599, will just ban us to fetch `requires_grad` from intermediate TensorProxy. However, it won't fix the problem of correctly propagating it #1768.

Even currently, it is safe to query requires_grad from input TensorProxy (if you know which ones they are). Example, for parameters, we can rely on their `requires_grad` and #1838 tried that. But this fails if there is a reshape or cast on a parameter before (as we can't trust the requires_grad on the output of these operations). See the folllowing test for an example of this case

https://github.com/Lightning-AI/lightning-thunder/blob/f73bfa0eef857532beeaea44964b0bfd9926325a/thunder/tests/test_transformer_engine_executor.py#L264-L283
